akka-projection-cassandra/src/main/scala/akka/projection/cassandra/internal/CassandraProjectionImpl.scala

```diff
@@ -111,9 +121,11 @@ import akka.stream.scaladsl.Source
       case AtMostOnce =>
         source
           .mapAsync(parallelism = 1) {
-            case (offset, envelope) => offsetStore.saveOffset(projectionId, offset).map(_ => offset -> envelope)
+            case (offset, envelope) =>
+              offsetStore
+                .saveOffset(projectionId, offset)
+                .flatMap(_ => applyUserRecovery(envelope, () => handler.process(envelope)))
```

Comment:
It was wrong because the async boundary may cause more than one offset to be saved before processing